packages/loader/driver-utils/src/runWithRetry.ts

```diff
@@ -30,40 +30,77 @@ export interface IProgress {
      * as well as information provided by service (like 429 error asking to wait for some time before retry)
      * @param error - error object returned from the call.
      */
     onRetry?(delayInMs: number, error: any): void;
 }
 
-export async function runWithRetry<T>(
+export interface IProgress2 {
+    /**
+     * Abort signal used to cancel operation
+     * Note that most of the layers do not use this signal yet. We need to change that over time.
+     * Please consult with API documentation / implementation.
+     * Note that  number of layers may not check this signal while holding this request in a queue,
+     * so it may take a while it takes effect. This can be improved in the future.
+     * Layers in question are:
+     *    - driver (RateLimiter)
+     *    - runWithRetry
+     */
+    cancel?: AbortSignal;
+
+    /**
+     * Called whenever api returns cancellable error and the call is going to be retried.
+     * @param delayInMs - delay before next retry. This value will depend on internal back-off logic,
+     * as well as information provided by service (like 429 error asking to wait for some time before retry)
+     * @param error - error object returned from the call.
+     * @returns - true to continue retrying, false to abort.
+     */
+    onRetry?(delayInMs: number, error: unknown): boolean;
+}
+
+export type RunResult<T> = {
+    status: "succeeded";
+    result: T;
+} | {
+    status: "failed";
+    error: unknown;
+} | {
+    status: "aborted";
+};
+
+export async function runWithRetry2<T>(
     api: (cancel?: AbortSignal) => Promise<T>,
     fetchCallName: string,
     logger: ITelemetryLogger,
-    progress: IProgress,
-): Promise<T> {
-    let result: T | undefined;
-    let success = false;
+    progress: IProgress2,
+): Promise<RunResult<T>> {
+    let successResult: RunResult<T> | undefined;
     let retryAfterMs = 1000; // has to be positive!
     let numRetries = 0;
     const startTime = performance.now();
     let lastError: any;
     do {
         try {
-            result = await api(progress.cancel);
-            success = true;
-        } catch (err) {
+            const result = await api(progress.cancel);
+            successResult = {
+                status: "succeeded",
+                result,
+            };
+        } catch (error) {
             // If it is not retriable, then just throw the error.
-            if (!canRetryOnError(err)) {
+            if (!canRetryOnError(error)) {
                 logger.sendTelemetryEvent({
                     eventName: `${fetchCallName}_cancel`,
                     retry: numRetries,
                     duration: performance.now() - startTime,
                     fetchCallName,
-                }, err);
-                throw err;
+                }, error);
+                return {
+                    status: "failed",
+                    error,
+                };
             }
 
             if (progress.cancel?.aborted === true) {
```

Comment:
i would move this out of the catch, and put it first thin in the loop, this will allow aborting in the onRetry handler

Reply:
this also feels more natural, as retry failure will supersede abort failure, and success will supersede both

Reply:
really just want to avoid multiple abort paths